File: windows/QuotaArc/Providers/CredentialCache.cs
namespace QuotaArc.Providers;

internal sealed class CredentialCache<T>
{
    private readonly object _gate = new();
    private T? _stored;
    private DateTime? _attemptedStamp;
    private DateTime? _attemptedAt;
    private Exception? _lastError;
    private readonly Func<T, bool> _isExpired;
    private readonly double _recheckAfter;
    private readonly double _retryAfterFailure;
    private readonly Func<DateTime> _now;

    public CredentialCache(
        Func<T, bool> isExpired,
        double recheckAfter = 5 * 60,
        double retryAfterFailure = 5 * 60,
        Func<DateTime>? now = null)
    {
        _isExpired = isExpired;
        _recheckAfter = recheckAfter;
        _retryAfterFailure = retryAfterFailure;
        _now = now ?? (() => DateTime.Now);
    }

    public T Value(Func<DateTime?> itemModifiedAt, Func<T> reload)
    {
        T? held;
        DateTime? askedStamp;
        DateTime? askedAt;
        Exception? failure;
        lock (_gate)
        {
            held = _stored;
            askedStamp = _attemptedStamp;
            askedAt = _attemptedAt;
            failure = _lastError;
        }

        if (held is not null && !_isExpired(held)) return held;

        var current = itemModifiedAt();
        var alreadyAsked = false;
        if (current is { } cur && askedStamp is { } stamp)
            alreadyAsked = cur == stamp;
        else if (askedAt is { } at)
        {
            var wait = failure is null ? _recheckAfter : _retryAfterFailure;
            alreadyAsked = (_now() - at).TotalSeconds < wait;
        }

        if (alreadyAsked)
        {
            if (held is not null) return held;
            if (failure is not null) throw failure;
        }

        try
        {
            var fresh = reload();
            lock (_gate)
            {
                _stored = fresh;
                _attemptedStamp = current;
                _attemptedAt = _now();
                _lastError = null;
            }
            return fresh;
        }
        catch (Exception ex)
        {
            lock (_gate)
            {
                _attemptedStamp = current;
                _attemptedAt = _now();
                _lastError = ex;
            }
            throw;
        }
    }

    public void Forget()
    {
        lock (_gate)
        {
            _stored = default;
            _attemptedStamp = null;
            _attemptedAt = null;
            _lastError = null;
        }
    }
}

File: Sources/Providers/CredentialCache.swift
import Foundation

/// Holds a borrowed credential so the keychain is read as rarely as possible.
///
/// Every read of the *data* is a chance for macOS to interrupt someone. The ACL
/// on another app's item does not list Quota Arc, so the first read prompts —
/// and if the answer was **Allow** rather than **Always Allow**, so does every
/// one after it.
///
/// The rule is not "hold it while it is valid". That was the rule, and it is
/// what made the app ask over and over: once a token aged out, the cache
/// refused to serve it and *every* caller went back to the keychain — once a
/// minute, all night, for a token that could not change until the owning app
/// next ran. Re-reading an unchanged item cannot produce a different answer; it
/// can only produce another dialogue.
///
/// So the question is whether the item has **changed**, which
/// `KeychainItem.modifiedAt` answers without touching the data and without
/// prompting. Expiry is left to the caller, who is the one that knows what an
/// expired token means.
final class CredentialCache<Credential>: @unchecked Sendable {
    private let lock = NSLock()
    private var stored: Credential?
    /// The item's modification date the last time we *asked* about it — on a
    /// refusal as much as on a success. Recording it only on success was what
    /// let a refused read be retried on every tick: it had no memory of having
    /// already asked about that version of the item.
    private var attemptedStamp: Date?
    private var attemptedAt: Date?
    /// Why the last attempt failed, kept so a caller can be given the same
    /// answer without macOS being asked again.
    private var lastError: Error?

    private let isExpired: (Credential) -> Bool
    /// How long to sit with what we have when the probe cannot tell us whether
    /// the item moved. Long enough not to nag, short enough that a rotation is
    /// picked up while you are still looking at the notch.
    private let recheckAfter: TimeInterval
    /// How long to leave macOS alone after it has said no, where there is no
    /// probe to be more precise with.
    private let retryAfterFailure: TimeInterval
    private let now: () -> Date

    init(recheckAfter: TimeInterval = 5 * 60,
         retryAfterFailure: TimeInterval = 5 * 60,
         now: @escaping () -> Date = Date.init,
         isExpired: @escaping (Credential) -> Bool) {
        self.recheckAfter = recheckAfter
        self.retryAfterFailure = retryAfterFailure
        self.now = now
        self.isExpired = isExpired
    }

    /// The held credential, or a fresh one when the item behind it has moved.
    ///
    /// `itemModifiedAt` must not read the item's data — the whole point is that
    /// it can be called freely.
    ///
    /// `reload` runs outside the lock: it can block on a keychain prompt, and
    /// holding a lock across a modal dialogue would stall every other caller
    /// behind it.
    func value(itemModifiedAt: () -> Date? = { nil },
               reload: () throws -> Credential) throws -> Credential {
        lock.lock()
        let held = stored
        let askedStamp = attemptedStamp
        let askedAt = attemptedAt
        let failure = lastError
        lock.unlock()

        // Still good: nothing to decide.
        if let held, !isExpired(held) { return held }

        // Expired, or nothing held at all. Ask the cheap question first.
        let current = itemModifiedAt()

        // Have we already put this exact version of the item to macOS?
        let alreadyAsked: Bool
        if let current, let askedStamp {
            // The probe settles it outright: the same item can only give the
            // same answer, whether that answer was a token or a refusal.
            alreadyAsked = current == askedStamp
        } else if let askedAt {
            // Nothing to compare against, so wait it out instead — and wait
            // longer after a refusal, because a refusal retried on a timer is
            // precisely what this exists to stop.
            let wait = failure == nil ? recheckAfter : retryAfterFailure
            alreadyAsked = now().timeIntervalSince(askedAt) < wait
        } else {
            alreadyAsked = false
        }

        if alreadyAsked {
            if let held { return held }
            // Nothing held and macOS has already refused this item: give back
            // the same answer rather than raising the same dialogue again.
            if let failure { throw failure }
        }

        do {
            let fresh = try reload()
            lock.lock()
            stored = fresh
            attemptedStamp = current
            attemptedAt = now()
            lastError = nil
            lock.unlock()
            return fresh
        } catch {
            lock.lock()
            // The attempt is recorded, the credential is not: a failure must
            // never be handed back as if it were one.
            attemptedStamp = current
            attemptedAt = now()
            lastError = error
            lock.unlock()
            throw error
        }
    }

    /// What is already in hand, and nothing more.
    ///
    /// For a caller that would like the credential's details but must not be
    /// the reason a dialogue appears. The settings row is exactly that: it is
    /// rebuilt every time the window renders, and it has no business asking
    /// macOS for a secret in order to print a plan name.
    var held: Credential? {
        lock.lock(); defer { lock.unlock() }
        return stored
    }

    /// Drop what is held, so the next read goes to the keychain for real.
    ///
    /// Called when the server rejects the credential — that is the one signal
    /// the copy in hand is wrong despite not having expired, which is exactly
    /// what happens when someone signs into a different account — and by
    /// "Allow access…", where raising the dialogue again *is* the point.
    func forget() {
        lock.lock()
        stored = nil
        attemptedStamp = nil
        attemptedAt = nil
        lastError = nil
        lock.unlock()
    }
}
